logging/logging.go

```diff
@@ -72,25 +95,76 @@ func GetLogLevel() int {
 	return LogDiscard
 }
 
-// CustomizeLogging changes logging format
-func CustomizeLogging(loggingFormat string, serviceName string) {
-	log.SetOutput(os.Stderr)
-	log.SetFormatter(logFormatterFor(loggingFormat, serviceName))
+// CustomizeBuilder allows to customize logging process
+type CustomizeBuilder struct {
+	writer        io.Writer
+	serviceName   string
+	loggingFormat string
+	hooks         []FormatterHook
+}
 
-	log.Debugf("Changed logging format to %s", loggingFormat)
+// Customize is a global function for logging customization.
+// Example of usage: Customize().SetServiceName(...).SetFormat(...).SetOutput(...).Complete()
+func Customize() *CustomizeBuilder {
+	return &CustomizeBuilder{}
 }
 
-func logFormatterFor(loggingFormat string, serviceName string) log.Formatter {
-	loggingFormat = strings.ToLower(loggingFormat)
+// SetOutput specifies where logs should be written (stderr, file, etc.)
+func (c *CustomizeBuilder) SetOutput(w io.Writer) *CustomizeBuilder {
+	c.writer = w
+	return c
+}
 
-	if loggingFormat == "json" {
-		return JSONFormatter(log.Fields{FieldKeyProduct: serviceName})
+// SetServiceName specifies global name of service that produces logs
+func (c *CustomizeBuilder) SetServiceName(serviceName string) *CustomizeBuilder {
+	c.serviceName = serviceName
+	return c
+}
+
+// SetFormat specifies actual format
+func (c *CustomizeBuilder) SetFormat(loggingFormat string) *CustomizeBuilder {
+	c.loggingFormat = loggingFormat
+	return c
+}
 
-	} else if loggingFormat == "cef" {
-		return CEFFormatter(log.Fields{FieldKeyProduct: serviceName})
+// SetHooks specifies additional pre-/post-processing actions with log entries
+func (c *CustomizeBuilder) SetHooks(hooks []FormatterHook) *CustomizeBuilder {
+	c.hooks = hooks
+	return c
+}
+
+// Complete finishes logging customization.
+// For default logging use Customize().Complete()
+func (c *CustomizeBuilder) Complete() {
+	if c.writer == nil {
+		c.writer = os.Stderr
+	}
+	if c.loggingFormat == "" {
+		c.loggingFormat = LoggingFormatPlaintext
 	}
+	/* We do not check hooks field, since it can be nil (standard log entry processing) */
+	log.SetOutput(c.writer)
+	log.SetFormatter(logFormatterFor(c.loggingFormat, c.serviceName, c.hooks))
+	log.Debugf("Changed logging format to %s", c.loggingFormat)
+}
 
-	return TextFormatter()
+/*
+// CustomizeLogging changes logging format
+func CustomizeLogging(loggingFormat string, serviceName string) {
+	log.SetOutput(os.Stderr)
+	log.SetFormatter(logFormatterFor(loggingFormat, serviceName, nil))
+	log.Debugf("Changed logging format to %s", loggingFormat)
+}*/
```

Comment:
Please delete the code if it’s not needed anymore, don’t leave it commented out.

If we leave it like this then no one will remember about it when we merge cop into master, and then there will be questions why a merge commit added this.